lib/restforce/db/attacher.rb

```diff
@@ -0,0 +1,76 @@
+module Restforce
+
+  module DB
+
+    # Restforce::DB::Attacher is responsible for cleaning up any orphaned
+    # upserted Salesforce records with the database records that they were
+    # originally supposed to be attached to. This allows us to successfully
+    # handle cases where a request timeout or partial failure would otherwise
+    # leave an upserted Salesforce record stranded without a related database
+    # record.
+    class Attacher < Task
+
+      # Public: Run the re-attachment process for any unsynchronized Salesforce
+      # records which have an external upsert ID.
+      #
+      # Returns nothing.
+      def run(*_)
+        return if @mapping.strategy.passive?
+
+        @runner.run(@mapping) do |run|
+          run.salesforce_instances.each { |instance| attach(instance) }
+        end
+      end
+
+      private
+
+      # Internal: Attach the passed Salesforce instance to a related database
+      # record.
+      #
+      # instance - A Restforce::DB::Instances::Salesforce.
+      #
+      # Returns nothing.
+      def attach(instance)
+        synchronization_id = instance.record.SynchronizationId__c
+
+        return unless synchronization_id
+        return unless valid_model?(synchronization_id)
+
+        # If the instance is already synchronized, then we just want to wipe the
+        # Synchronization ID and proceed to the next instance.
+        if instance.synced?
+          instance.update!("SynchronizationId__c" => nil)
+          return
+        end
+
+        record_id = synchronization_id.split("::").last
```

Comment:
I still don't love that we're using this format, and that the knowledge about the format is spreading throughout the codebase. :/